internal/xds/xdsdependencymanager/xds_dependency_manager.go

```diff
@@ -0,0 +1,241 @@
+	// resource. This includes cases where:
+	//  - The listener or route resource watcher reports a resource error.
+	//  - The received listener resource is a socket listener, not an API listener - TODO :  This is not yet implemented, tracked here #8114
+	//  - The received route configuration does not contain a virtual host
+	//  matching the channel's default authority.
+	OnError(error)
+}
+
+func (m *DependencyManager) maybeSendUpdate() {
+	if m.logger.V(2) {
+		m.logger.Infof("Sending update to watcher: Listener: %v, RouteConfig: %v", pretty.ToJSON(m.currentListenerUpdate), pretty.ToJSON(m.currentRouteConfig))
+	}
+	m.watcher.OnUpdate(xdsresource.XDSConfig{
+		Listener:    *m.currentListenerUpdate,
+		RouteConfig: m.currentRouteConfig,
+		VirtualHost: m.currentVirtualHost,
+	})
+}
+
+func (m *DependencyManager) applyRouteConfigUpdate(update xdsresource.RouteConfigUpdate) {
+	matchVh := xdsresource.FindBestMatchingVirtualHost(m.dataplaneAuthority, update.VirtualHosts)
+	if matchVh == nil {
+		m.watcher.OnError(status.Errorf(codes.Unavailable, "Could not find VirtualHost for %q", m.dataplaneAuthority))
+		return
+	}
+	m.currentRouteConfig = update
+	m.currentVirtualHost = matchVh
+	m.maybeSendUpdate()
+}
+
+// Only executed in the context of a serializer callback.
+func (m *DependencyManager) onListenerResourceUpdate(update *xdsresource.ListenerUpdate) {
+	if m.logger.V(2) {
+		m.logger.Infof("Received update for Listener resource %q: %v", m.ldsResourceName, pretty.ToJSON(update))
+	}
+
+	m.currentListenerUpdate = update
+
+	if update.InlineRouteConfig != nil {
+		// If there was a previous route config watcher because of a non-inline
+		// route configuration, cancel it.
+		m.rdsResourceName = ""
+		if m.routeConfigWatcher != nil {
+			m.routeConfigWatcher.stop()
+			m.routeConfigWatcher = nil
+		}
+
+		m.applyRouteConfigUpdate(*update.InlineRouteConfig)
+		return
+	}
+
+	// We get here only if there was no inline route configuration.
+
+	// If the route config name has not changed, send an update with existing
+	// route configuration and the newly received listener configuration.
+	if m.rdsResourceName == update.RouteConfigName {
+		m.maybeSendUpdate()
+		return
+	}
+
+	// If the route config name has changed, cancel the old watcher and start a
+	// new one. At this point, since we have not yet resolved the new route
+	// config name, we don't send an update to the channel, and therefore
+	// continue using the old route configuration (if received) until the new
+	// one is received.
+	m.rdsResourceName = update.RouteConfigName
+	if m.routeConfigWatcher != nil {
+		m.routeConfigWatcher.stop()
+		m.currentVirtualHost = nil
+	}
+	m.routeConfigWatcher = newRouteConfigWatcher(m.rdsResourceName, m)
+}
+
+// Only executed in the context of a serializer callback.
+func (m *DependencyManager) onListenerResourceError(err error) {
+	if m.logger.V(2) {
+		m.logger.Infof("Received resource error for Listener resource %q: %v", m.ldsResourceName, err)
```

Comment:
I believe we have some code in the xDS client to ensure that the returned errors contain the xDS node ID. Could you please ensure that that property still holds. Thanks.